src/ring_theory/graded_algebra/homogeneous_ideal.lean

```diff
@@ -0,0 +1,427 @@
+  I.is_homogeneous 𝒜 ↔ ∀ i, (I : set A) ⊆ graded_algebra.proj 𝒜 i ⁻¹' I :=
+iff.rfl
+
+lemma ideal.is_homogeneous_iff_subset_Inter :
+  I.is_homogeneous 𝒜 ↔ (I : set A) ⊆ ⋂ i, graded_algebra.proj 𝒜 i ⁻¹' ↑I :=
+subset_Inter_iff.symm
+
+lemma ideal.is_homogeneous.exists_iff_eq_span :
+  (∃ (S : set (homogeneous_submonoid 𝒜)), I = ideal.span (coe '' S)) ↔
+    I = I.homogeneous_core' 𝒜 :=
+(set.image_preimage.compose (submodule.gi _ _).gc).exists_eq_l _
+
+lemma mul_homogeneous_element_mem_of_mem
+  {I : ideal A} (r x : A) (hx₁ : is_homogeneous 𝒜 x) (hx₂ : x ∈ I) (j : ι) :
+  graded_algebra.proj 𝒜 j (r * x) ∈ I :=
+begin
+  letI : Π (i : ι) (x : 𝒜 i), decidable (x ≠ 0) := λ _ _, classical.dec _,
+  rw [←graded_algebra.sum_support_decompose 𝒜 r, finset.sum_mul, linear_map.map_sum],
+  apply ideal.sum_mem,
+  intros k hk,
+  obtain ⟨i, hi⟩ := hx₁,
+  have mem₁ : (graded_algebra.decompose 𝒜 r k : A) * x ∈ 𝒜 (k + i) := graded_monoid.mul_mem
+    (submodule.coe_mem _) hi,
+  rw [graded_algebra.proj_apply, graded_algebra.decompose_of_mem 𝒜 mem₁,
+    coe_of_submodule_apply 𝒜, submodule.coe_mk],
+  split_ifs,
+  { exact I.mul_mem_left _ hx₂ },
+  { exact I.zero_mem },
+end
+
+lemma ideal.is_homogeneous.homogeneous_core : (I.homogeneous_core' 𝒜).is_homogeneous 𝒜 :=
+begin
+  rintros i r hr,
+  rw [ideal.homogeneous_core', ideal.span, finsupp.span_eq_range_total] at hr,
+  rw linear_map.mem_range at hr,
+  obtain ⟨s, rfl⟩ := hr,
+  rw [←graded_algebra.proj_apply, finsupp.total_apply, finsupp.sum, linear_map.map_sum],
+  refine ideal.sum_mem _ _,
+  rintros z hz1,
+  rw [smul_eq_mul],
+  refine mul_homogeneous_element_mem_of_mem 𝒜 (s z) z _ _ i,
+  { rcases z with ⟨z, hz2⟩,
+    rw subtype.image_preimage_coe at hz2,
+    exact hz2.2, },
+  { exact ideal.subset_span z.2 },
+end
+
+/--Bundled version of `homogeneous_core`, i.e. given ideal `I`, the homogeneous ideal
+`homogeneous_core' 𝒜 I`-/
+abbreviation ideal.homogeneous_core : homogeneous_ideal 𝒜 :=
+⟨ideal.homogeneous_core' 𝒜 I, ideal.is_homogeneous.homogeneous_core 𝒜 I⟩
+
+variables {𝒜 I}
+
+lemma ideal.is_homogeneous.homogeneous_core_eq_self (h : I.is_homogeneous 𝒜) :
+  I.homogeneous_core' 𝒜 = I :=
+begin
+  symmetry,
+  apply le_antisymm _ (I.homogeneous_core'_le_ideal 𝒜),
+  intros x hx,
+  letI : Π (i : ι) (x : 𝒜 i), decidable (x ≠ 0) := λ _ _, classical.dec _,
+  rw ←graded_algebra.sum_support_decompose 𝒜 x,
+  exact ideal.sum_mem _ (λ j hj, ideal.subset_span ⟨⟨_, is_homogeneous_coe _⟩, h _ hx, rfl⟩),
+end
+
+variables (𝒜 I)
+
+lemma ideal.is_homogeneous.iff_eq :
+  I.is_homogeneous 𝒜 ↔ I.homogeneous_core' 𝒜 = I:=
+⟨ λ hI, hI.homogeneous_core_eq_self,
+  λ hI, hI ▸ ideal.is_homogeneous.homogeneous_core 𝒜 I ⟩
+
+lemma ideal.is_homogeneous.iff_exists :
+  I.is_homogeneous 𝒜 ↔ ∃ (S : set (homogeneous_submonoid 𝒜)), I = ideal.span (coe '' S) :=
+begin
+  rw [ideal.is_homogeneous.exists_iff_eq_span, ideal.is_homogeneous.iff_eq],
+  exact ⟨λ h, h.symm, λ h, h.symm⟩
+end
```

Comment:
So we can remove the confusingly named `ideal.is_homogeneous.exists_iff_eq_span` above:
```suggestion
lemma ideal.is_homogeneous.iff_exists :
  I.is_homogeneous 𝒜 ↔ ∃ (S : set (homogeneous_submonoid 𝒜)), I = ideal.span (coe '' S) :=
begin
  rw [ideal.is_homogeneous.iff_eq, eq_comm],
  exact ((set.image_preimage.compose (submodule.gi _ _).gc).exists_eq_l _).symm,
end
```